File: Commons/Math/src/main/java/com/greentree/commons/math/vector/Vector3f.java
package com.greentree.commons.math.vector;

import org.joml.Matrix3f;

public class Vector3f extends AbstractVector3f {

	public float x, y, z;
	
	private static final long serialVersionUID = 1L;

	public Vector3f() {
	}

	public Vector3f(AbstractVector2f f) {
		super(f);
	}
	public Vector3f(AbstractVector2f f, float z) {
		super(f, z);
	}
	public Vector3f(AbstractVector3f vec) {
		super(vec);
	}
	public Vector3f(float f) {
		super(f);
	}

	public Vector3f(float x, float y, float z) {
		super(x, y, z);
	}
	@Override
	public Vector3f add(float x, float y, float z) {
		return add(x, y, z, this);
	}
	@Override
	public Vector3f cross(AbstractVector3f vec) {
		return cross(vec, this);
	}
	public Vector3f mul(AbstractVector3f v) {
		return mul(v, this);
	}

	@Override
	public Vector3f mul(float f) {
		return mul(f, this);
	}

	@Override
	public Vector3f mul(Matrix3f mat) {
		return mul(mat, this);
	}

	@Override
	public Vector3f normalize() {
		return normalize(1, this);
	}

	@Override
	public void x(float x) {
		this.x = x;
	}

	@Override
	public void y(float y) {
		this.y = y;
	}

	@Override
	public void z(float z) {
		this.z = z;
	}

	@Override
	public float x() {
		return x;
	}

	@Override
	public float y() {
		return y;
	}

	@Override
	public float z() {
		return z;
	}

}

File: Commons/Math/src/main/java/com/greentree/commons/math/Vector3f.kt
package com.greentree.commons.math

import org.joml.Vector2f
import org.joml.Vector2fc
import org.joml.Vector3f
import org.joml.Vector3fc

val Vector3fc.x
	get() = x()
val Vector3fc.y
	get() = y()
val Vector3fc.z
	get() = z()

fun vec3f(xy: Vector2fc, z: Float) = vec3f(xy.x, xy.y, z)
fun vec3f(x: Float, y: Float, z: Float) = Vector3f(x, y, z)

fun Vector3fc.cross(other: Vector3fc) =
	Vector3f(y * other.z - z * other.y, z * other.x - x * other.z, x * other.y - y * other.x)

fun Vector3fc.projection(normal: Vector3fc): Vector2f {
	val a = normal.x()
	val b = normal.y()
	val c = normal.z()
	var xAxis: Vector3fc
	var yAxis: Vector3fc = vec3f(a, b, c).normalize()
	xAxis = vec3f(b, -a, 0f)
	if(xAxis.lengthSquared() < Mathf.EPS)
		xAxis = vec3f(0f, c, -b)
	yAxis = yAxis.cross(xAxis)
	val x = dot(xAxis)
	val y = dot(yAxis)
	return vec2f(x, y)
}